From: Samuel Thibault <samuel.thibault@ens-lyon.org>
Subject: [PATCH] The critical section lock _can_ be held in these place.

At least since hurd_thread_cancel can be called by another thread and lock our
critical lock.

http://bugs.debian.org/46859

“
Thomas suggested that there is no need to take the critical section
lock.  I believe that taking the critical section lock is necessary to
prevent the target thread from entering a signal handler.  Roland will
look into the problem.
”

Taking the critical section lock makes these assertions bogus.

It happens that hurd_thread_cancel is only called from libports and inside
/hurd/term so this is rare in practice.

A reproducer can be found here:

http://lists.gnu.org/archive/html/bug-hurd/2014-05/msg00025.html

2006-08-05  Samuel Thibault  <samuel.thibault@ens-lyon.org>

       * hurd/thread-cancel.c (hurd_thread_cancel): Do not assert that
       `&ss->critical_section_lock' is unlocked.
       * sysdeps/mach/hurd/jmp-unwind.c (_longjmp_unwind): Likewise, and take
       critical section lock before taking the sigstate lock.
       * sysdeps/mach/hurd/spawni.c (__spawni): Likewise.

---
 hurd/hurdexec.c                | 1 -
 hurd/thread-cancel.c           | 2 --
 sysdeps/mach/hurd/jmp-unwind.c | 3 +--
 sysdeps/mach/hurd/spawni.c     | 1 -
 4 files changed, 1 insertion(+), 6 deletions(-)

Index: glibc-2.28/hurd/hurdexec.c
===================================================================
--- glibc-2.28.orig/hurd/hurdexec.c
+++ glibc-2.28/hurd/hurdexec.c
@@ -123,7 +123,6 @@ _hurd_exec_paths (task_t task, file_t fi
 
   ss = _hurd_self_sigstate ();
 
-  assert (! __spin_lock_locked (&ss->critical_section_lock));
   __spin_lock (&ss->critical_section_lock);
 
   __spin_lock (&ss->lock);
Index: glibc-2.28/hurd/thread-cancel.c
===================================================================
--- glibc-2.28.orig/hurd/thread-cancel.c
+++ glibc-2.28/hurd/thread-cancel.c
@@ -51,7 +51,6 @@ hurd_thread_cancel (thread_t thread)
       return 0;
     }
 
-  assert (! __spin_lock_locked (&ss->critical_section_lock));
   __spin_lock (&ss->critical_section_lock);
   __spin_lock (&ss->lock);
   err = __thread_suspend (thread);
@@ -91,7 +90,6 @@ hurd_check_cancel (void)
   int cancel;
 
   __spin_lock (&ss->lock);
-  assert (! __spin_lock_locked (&ss->critical_section_lock));
   cancel = ss->cancel;
   ss->cancel = 0;
   __spin_unlock (&ss->lock);
Index: glibc-2.28/sysdeps/mach/hurd/jmp-unwind.c
===================================================================
--- glibc-2.28.orig/sysdeps/mach/hurd/jmp-unwind.c
+++ glibc-2.28/sysdeps/mach/hurd/jmp-unwind.c
@@ -49,9 +49,8 @@ _longjmp_unwind (jmp_buf env, int val)
 
   /* All access to SS->active_resources must take place inside a critical
      section where signal handlers cannot run.  */
-  __spin_lock (&ss->lock);
-  assert (! __spin_lock_locked (&ss->critical_section_lock));
   __spin_lock (&ss->critical_section_lock);
+  __spin_lock (&ss->lock);
 
   /* Remove local signal preemptors being unwound past.  */
   while (ss->preemptors &&
Index: glibc-2.28/sysdeps/mach/hurd/spawni.c
===================================================================
--- glibc-2.28.orig/sysdeps/mach/hurd/spawni.c
+++ glibc-2.28/sysdeps/mach/hurd/spawni.c
@@ -239,7 +239,6 @@ __spawni (pid_t *pid, const char *file,
 
   ss = _hurd_self_sigstate ();
 
-  assert (! __spin_lock_locked (&ss->critical_section_lock));
   __spin_lock (&ss->critical_section_lock);
 
   __spin_lock (&ss->lock);
